x/liquidity/keeper/migration.go

```diff
@@ -1,22 +1,164 @@
 	return v043.MigrateStore(ctx, m.keeper.storeKey)
 }
+
+// Migrate2to3 migrates from version 2 to 3.
+func (m Migrator) Migrate2to3(ctx sdk.Context) error {
+	return SafeForceWithdrawal(ctx, m.keeper, m.bankKeeper, m.accountKeeper)
+}
+
+// SafeForceWithdrawal call ForceWithdrawal safely by recover, cached ctx
+func SafeForceWithdrawal(ctx sdk.Context, keeper Keeper, bankKeeper liquiditytypes.BankKeeper, accountKeeper liquiditytypes.AccountKeeper) error {
+	broken := false
+	defer func() {
+		if r := recover(); r != nil {
+			broken = true
+		}
+	}()
+
+	cachedCtx, writeCache := ctx.CacheContext()
+	err := ForceWithdrawal(cachedCtx, keeper, bankKeeper, accountKeeper)
+	if err == nil && !broken {
+		writeCache()
+	}
+	return nil
+}
+
+// ForceWithdrawal Forcefully withdraw pool token holders once migration
+func ForceWithdrawal(ctx sdk.Context, keeper Keeper, bankKeeper liquiditytypes.BankKeeper, accountKeeper liquiditytypes.AccountKeeper) error {
+	logger := keeper.Logger(ctx)
+	poolByPoolCoinDenom := map[string]liquiditytypes.Pool{} // PoolCoinDenom => Pool
+	for _, pool := range keeper.GetAllPools(ctx) {
+		poolByPoolCoinDenom[pool.PoolCoinDenom] = pool
+	}
+
+	// Iterate over all the balances of all accounts and find the accounts that hold poolxxx... coin.
+	// Unless it is pool reserve account, forcefully withdraw their pool coin and transfer
+	// the corresponding amount of respective reserve coins back to their accounts.
+	// Lastly, burn pool coins that are withdrawn.
+	accMap := map[string]authtypes.AccountI{}
+	accList := []string{}
+	bankKeeper.IterateAllBalances(ctx, func(address sdk.AccAddress, coin sdk.Coin) (stop bool) {
+		if strings.HasPrefix(coin.Denom, "pool") {
+			pool := poolByPoolCoinDenom[coin.Denom]
+			if _, ok := accMap[address.String()]; !ok {
+				accMap[address.String()] = accountKeeper.GetAccount(ctx, address)
+				accList = append(accList, address.String())
+			}
+
+			// Skip pool reserve accounts
+			acc := accMap[address.String()]
+			if acc.GetSequence() != 0 || acc.GetPubKey() != nil {
+				if _, err := keeper.WithdrawWithinBatch(ctx, &liquiditytypes.MsgWithdrawWithinBatch{
+					WithdrawerAddress: address.String(),
+					PoolId:            pool.GetId(),
+					PoolCoin:          coin,
+				}); err != nil {
+					logger.Debug(
+						"failed force withdrawal",
+						"withdrawer", address.String(),
+						"poolcoin", coin,
+						"error", err,
+					)
+				}
+			}
+		}
+		return false
+	})
+
+	// Execute batch manually
+	keeper.ExecutePoolBatches(ctx)
+	keeper.DeleteAndInitPoolBatches(ctx)
+
+	// iterating and withdraw again if there is any pool coin left due to decimal error
+	for _, address := range accList {
+		acc := accMap[address]
+		balances := keeper.bankKeeper.GetAllBalances(ctx, acc.GetAddress())
+		for _, coin := range balances {
+			if strings.HasPrefix(coin.Denom, "pool") {
+				// Skip pool reserve accounts
+				if acc.GetSequence() != 0 || acc.GetPubKey() != nil {
```

Comment:
nitpicking:  account sequences can’t be incremented without a signing key afaiu. So this condition might be simplified to: `acc.GetPubKey() != nil`